### PR TITLE
all: migrate gradle publish from maven to use maven-publish plugin

### PR DESCRIPTION
**Pull request by creamsoup**
You can see sample snapshot repo at http://jihuncho0.svl.corp.google.com:8081/#browse/browse:jihun_staging

### CODE REVIEW

File: netty/build.gradle [hunk 1]
```diff
@@ -33,9 +33,3 @@ test {
     // Allow testing Jetty ALPN in TlsTest
     jvmArgs "-javaagent:" + configurations.alpnagent.asPath
 }
-
-jmh {
```
Comment (creamsoup) [inline netty/build.gradle]: this caused an error during :grpc-netty:animalsnifferJmh task.
error is about "Failed to check signatures"

**Comment by creamsoup** [comment]
some differences are 
1. grpc-gae-interop-testing-jdk8-1.0-SNAPSHOT.pom packaging "war" is removed.  (probably need to fix?)
2. runtime scope is changed to compile.
3. test dependencies are removed.

**Comment by ejona86** [comment]
grpc-gae-interop-testing-jdk8 should not be uploaded.

compile scope instead of runtime is unfortunate, but probably okay. I think I remember seeing some Gradle discussion about this. It might be fixed if we swapped to implementation and api dependencies.

File: build.gradle [hunk 2]
```diff
@@ -348,79 +342,101 @@ subprojects {
...
+                artifact sourcesJar
 
-            licenses {
-                license {
-                    name 'Apache 2.0'
-                    url 'https://opensource.org/licenses/Apache-2.0'
-                }
-            }
+                pom {
+                    name = "$project.group:$project.name".toString()
```
Comment (ejona86) [inline build.gradle]: Why is toString() necessary?

File: build.gradle [hunk 3]
```diff
@@ -348,79 +342,101 @@ subprojects {
...
+                        }
+                    }
+
+                    developers {
+                        developer {
+                            id = "grpc.io"
+                            name = "gRPC Contributors"
+                            email = "grpc-io@googlegroups.com"
+                            url = "https://grpc.io/"
+                            // https://issues.gradle.org/browse/GRADLE-2719
```
Comment (ejona86) [inline build.gradle]: This may now be out-of-date.

File: netty/shaded/build.gradle [hunk 4]
```diff
@@ -46,14 +46,6 @@ shadowJar {
     mergeServiceFiles()
 }
 
-// This is a hack to have shadow plugin modify the uploadArchives POM's
```
Comment (ejona86) [inline netty/shaded/build.gradle]: grpc-netty-shaded's pom doesn't look right. It shouldn't have dependencies on netty and tcnative.

File: settings.gradle [hunk 5]
```diff
@@ -1,44 +1,45 @@
 rootProject.name = "grpc"
-include ":grpc-core"
```
Comment (ejona86) [inline settings.gradle]: Why were these lines changed?

File: netty/build.gradle (hunk 1 above)
Comment (ejona86) [reply netty/build.gradle]: Check whether "./gradlew :grpc-netty:jmh"  runs. You don't need to do a full benchmark, but if a benchmark starts, it is probably fine.

File: build.gradle [hunk 6]
```diff
@@ -348,79 +342,101 @@ subprojects {
         from sourceSets.main.allSource
     }
 
-    artifacts { archives javadocJar, sourcesJar }
```
Comment (ejona86) [inline build.gradle]: Why is the artifacts block no longer used here, but we still use it other places (like compiler)?

File: compiler/build.gradle [hunk 7]
```diff
@@ -1,5 +1,6 @@
 apply plugin: "cpp"
 apply plugin: "com.google.protobuf"
+apply plugin: "java"
```
Comment (ejona86) [inline compiler/build.gradle]: This looks fishy.

File: netty/shaded/build.gradle [hunk 8]
```diff
@@ -46,13 +41,15 @@ shadowJar {
...
-// but that's broken anyway with install/uploadArchives
-uploadShadow.repositories.addAll(uploadArchives.repositories)
-// And then we use a further hack to share that POM with install
-install.repositories.mavenInstaller.pom = uploadArchives.repositories.mavenDeployer.pom
+publishing {
+    publications {
+        maven(MavenPublication) {
+            // Ideally swap to project.shadow.component(it) when it isn't broken for project deps
+            artifact shadowJar
+            // TODO: manually add deps
```
Comment (ejona86) [inline netty/shaded/build.gradle]: TODO: here and elsewhere

File: build.gradle [hunk 9]
```diff
@@ -348,79 +342,105 @@ subprojects {
...
+                            username = rootProject.ossrhUsername
+                            password = rootProject.ossrhPassword
+                        }
+                    }
+                    def releaseUrl = stagingUrl
+                    def snapshotUrl = 'https://oss.sonatype.org/content/repositories/snapshots/'
+                    url = version.endsWith('SNAPSHOT') ? snapshotUrl : releaseUrl
+                }
             }
+            mavenLocal()
```
Comment (ejona86) [inline build.gradle]: Remove (unnecessary)

File: build.gradle (hunk 6 above)
Comment (creamsoup) [reply build.gradle]: those are unused and removed.

File: build.gradle (hunk 2 above)
Comment (creamsoup) [reply build.gradle]: GString(Impl)? seems to have issue (there are few open issues and some are closed but still doesn't quite work)
manual string concat works without toString.

File: build.gradle (hunk 3 above)
Comment (creamsoup) [reply build.gradle]: yes. the issue is marked as closed. removing the note.

File: build.gradle [hunk 10]
```diff
@@ -348,79 +342,101 @@ subprojects {
...
-            "grpc-protobuf",
-            "grpc-protobuf-lite",
-            "grpc-protobuf-nano"
-        ])) {
-            def core = pom.dependencies.find {dep -> dep.artifactId == 'grpc-core'}
-            if (core != null) {
-                // Depend on specific version of grpc-core because internal package is unstable
-                core.version = "[" + core.version + "]"
+        repositories {
+            maven {
```
[the remark this answers is not in the document]
Comment (creamsoup) [reply build.gradle]: so publishToMavenLocal is not using the repositoryDir to specify
./gradlew publishToMavenLocal -Dmaven.repo.local=$MY_M2_HOME
or 
./gradlew publish -PrepositoryDir=$MY_M2_HOME

File: compiler/build.gradle (hunk 7 above)
Comment (creamsoup) [reply compiler/build.gradle]: oops

File: netty/build.gradle (hunk 1 above)
Comment (creamsoup) [reply netty/build.gradle]: it was broken (it maybe due to some hack in the previous version of PR). now it works as expected.

File: netty/shaded/build.gradle (hunk 4 above)
Comment (creamsoup) [reply netty/shaded/build.gradle]: discussed offline. NOTE for others: because shadow plugin is filtering project deps, we had to manually generate dependencies.

File: settings.gradle (hunk 5 above)
Comment (creamsoup) [reply settings.gradle]: reverted to reduce review complexity. i'll issue other PR for sorting the order.

**Comment by creamsoup** [comment]
TODO: i think some checks will fail (especially android), need to fix the some build scripts (may still need to install since android ones are not converted)
